neural_nets/activations/activations.py

```diff
@@ -232,3 +232,29 @@ def grad(self, x):
 
     def grad2(self, x):
         return np.where(x >= 0, np.zeros_like(x), np.exp(x) * self.alpha * self.scale)
+
+
+class Softmax(ActivationBase):
```

Comment:
The softmax gradient independent from a cross-entropy loss is a bit tricky - you have to make sure you do all pairwise partial derivatives. Your current `grad` is only computing the diagonal of the Jacobian.

I implemented the full gradient calculation for 2D input [here](https://github.com/ddbourgin/numpy-ml/blob/9aff8eb1cab1b46eabfe113f59806d91f16005ce/neural_nets/layers/layers.py#L1872), so I think this may not be strictly necessary.

Reply:
I will delete this part.